alibi_detect/od/backends/keops/knn.py

```diff
@@ -0,0 +1,25 @@
+import numpy as np
+import torch
+from pykeops.torch import LazyTensor
+
+
+def cdist(X, Y):
+    return ((X - Y)**2).sum(-1).sqrt()
+
+
+class KNNKeops:
+    def score(X, x_ref, k, kernel=None):
+        ensemble = isinstance(k, (np.ndarray, list, tuple))
+        X = torch.as_tensor(X)
+        X_keops = LazyTensor(X[:, None, :])
+        x_ref_keops = LazyTensor(x_ref[None, :, :])
+        K = -kernel(X_keops, x_ref_keops) if kernel else cdist(X_keops, x_ref_keops)
+        ks = np.array(k) if ensemble else np.array([k])
+        bot_k_inds = K.argKmin(np.max(ks), dim=1)
+        all_knn_dists = (X[:, None, :] - x_ref[bot_k_inds][:, ks-1, :]).norm(dim=2)
```

Comment:
Doesn't this defeat the `keops` speed up because you are basically doing the same as the distance computation here? You might be able to use `Kmin_argKmin` instead of `argKmin` in the line above and compute both the indices and distances within `keops` at once. Once you have those you can index into it for the ensemble.

Reply:
Ah, I see what you mean. I think I'd assumed that `len(bot_k_inds)` is much less than `len(x_ref)` and `len(X)` and hence `(X[:, None, :] - x_ref[bot_k_inds][:, ks-1, :]).norm(dim=2)` wouldn't need to use `keops`?

Reply:
I think (not 100% sure due to multiple possible k's) you can do this within `keops` using `Kmin_argKmin` which would be drastically faster (expanding dims and doing broadcasted operations is slow and memory intensive). If `Kmin_argKmin` does not suffice for multiple k's, then you could implement this as a custom formula using a generic reduction. Check [this page](https://www.kernel-operations.io/keops/python/api/pytorch/Genred_torch.html) for more info.